lib/web/app/handler.go

```diff
@@ -140,29 +141,63 @@ func (h *Handler) handleForward(w http.ResponseWriter, r *http.Request, session
 	return nil
 }
 
+// handleForwardError when the forwarder has an error during the `ServeHTTP` it
+// will calll this function. This handler will then renew the session in order
+// to get "fresh" app servers, and then will forwad the request to the newly
+// created session.
+func (h *Handler) handleForwardError(w http.ResponseWriter, req *http.Request, err error) {
+	if !trace.IsConnectionProblem(err) && err != io.EOF {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+		return
+	}
+
+	session, err := h.renewSession(req)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(http.StatusText(http.StatusInternalServerError)))
+		return
+	}
+
+	session.fwd.ServeHTTP(w, req)
+}
```

Comment:
Not really, the new session flow now only matches "healthy" servers, so if there isn't any, it will return an error on the renew the session function.

However, there can be a case where the server is unstable, meaning that at the session creation, it is healthy, but when the forward happens, it is not. So we'll have the loop you mentioned in this case, and the client will timeout its request.